### PR TITLE
Implement bst tree

### PR DESCRIPTION
**Pull request by danghai**
Hi, I implement the bst tree. `insert`, and `search` works now. But I am not sure why `size` method does not work: 
The error:  `AttributeError: 'NoneType' object has no attribute 'size' `
@goswami-rahul  @christianbender Could you help me?

### CODE REVIEW

File: tree/bst/bst.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            return False
+        elif data > self.data:     # Go to right root
+            return self.right.search(data)
+        else:                      # Go to left root
+            return self.left.search(data)
+    """
+        Get the number of elements
+        Using recursion. Complexity O(logN)
+    """
+    def size(self):
```
Comment (goswami-rahul) [inline tree/bst/bst.py]: these methods must be defined inside `BST` class, not inside `Node`.

**Comment by danghai** [comment]
@tratneshsingh Thank you.

File: tree/bst/bst.py (hunk 1 above)
Comment (danghai) [reply tree/bst/bst.py]: What do you mean?

**Comment by tratneshsingh** [comment]
For remove_max, think of a case where you remove the only node present in the tree. This shall give you a clue that you need to set the self.root = <>. So, should be the case for remove method.

**Comment by danghai** [comment]
@tratneshsingh Hmm I do not get what you mean. I think I need to rewrite these method under BST first.

**Comment by goswami-rahul** [comment]
@danghai I mean, class `Node` just represents a single node. It doesn't has any methods in itself. However, the methods like `size()` and `search()` does _not_ apply to _Node_, these apply to a _Tree_, such as our `BST`. Therefore, these entire methods should be moved to `BST` class.
Sorry, if you didn't understand what I mean, I will merge it and commit it myself :)

**Comment by goswami-rahul** [comment]
Even better! I will make an abstract class `Tree` implementing these methods, and BST or any other tree can be subclassed from there.

**Comment by danghai** [comment]
@goswami-rahul No I know what you mean :), I actually write these method under bst, However, I have struggle the node inside when I use recursion. I am going to rewrite them under bst now. 
Yeah, you help me to build them as a abstract class then

**Comment by danghai** [comment]
@goswami-rahul Ok, I have updated. However, `remove_max` does not work.

File: tree/bst/bst.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        else:                      # Go to right root
+            if root.right:         # If right root is a node
+                return self.recur_insert(root.right, data)
+            else:
+                root.right = Node(data)
+                return True
+
+    def remove(self, data):
+        if self.root:
+            return self.root.remove(data)
```
Comment (goswami-rahul) [inline tree/bst/bst.py]: There is no method as `remove` in class `Node`. Did you run your tests before committing?

**Comment by goswami-rahul** [comment]
@danghai Please understand that incorrect code is not acceptable. If there are even some slight bugs, it is better to not include that method for now. They will surely be implemented by any of us in the future.

**Comment by danghai** [comment]
@goswami-rahul , I do not implement the `remove` yet because the code `remove_max` does not work. I forget comment out

**Comment by goswami-rahul** [comment]
Thanks. Looks good.

**Pull request merged**